### PR TITLE
Trainer: set skip_logits for loss-only eval when liger enabled

### PR DESCRIPTION
**Pull request by AkshajKashyap**
Fixes #43039

## What does this PR do?
When `prediction_loss_only=True` during evaluation and `use_liger_kernel=True`, `Trainer.prediction_step` now passes `skip_logits=True` to the model forward if the forward signature supports it and labels are present.

This avoids materializing logits during loss-only eval and enables fused loss paths for implementations that use `skip_logits` (for example, Liger kernel integrations), which can reduce memory usage during evaluation.

## Implementation details
- Injects `skip_logits=True` into `inputs` only when:
  - `prediction_loss_only` is true
  - `use_liger_kernel` is enabled
  - `labels` are present
  - model forward accepts a `skip_logits` parameter (checked via signature)
- No behavior change when labels are missing, or when the model does not support `skip_logits`.

## Tests
Added CPU-only unit tests:
- `test_trainer_sets_skip_logits_for_loss_only_eval_when_liger_enabled`
- `test_trainer_does_not_set_skip_logits_when_no_labels_but_return_loss_true`

Run with:
`python -m pytest -q tests/trainer/test_skip_logits_eval.py -x`

## Code Agent Policy
- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the contributor guideline, Pull Request section?
- [x] Was this discussed/approved via a Github issue or the forum? Issue: #43039
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?
Tagging: @SunMarc

### CODE REVIEW

**Review by SunMarc — COMMENTED** [review]
did you see memory gain with this ?

**Comment by AkshajKashyap** [comment]
> did you see memory gain with this ?

Yep, I measured a clear peak GPU memory reduction on my RTX 3050 (4GB) using a small Llama model with Liger enabled.

Benchmark setup:
- Model: hf-internal-testing/tiny-random-LlamaForCausalLM
- batch=1, seq_len=1024, fp16
- Measured peak allocated GPU memory during loss-only eval

Results:
- baseline (forced skip_logits=False): 324.7 MB peak
- with this PR behavior (Trainer injects skip_logits=True when supported): 10.6 MB peak
- delta: 314.0 MB (~96.7%)

This matches the intent of the fix: in loss-only eval, skipping logits avoids materializing the logits tensor and enables fused loss paths for implementations that use skip_logits (like Liger integrations). Savings should generally increase with longer sequence length / larger vocab.

Appreciate your reply, and really just trying to help and make myself useful.

**Comment by AkshajKashyap** [comment]
> did you see memory gain with this ?

Quick follow-up since it’s been about a week and CI is still green: is this change directionally OK, or would you prefer a different guard / placement?

Happy to adjust quickly if you want this behind an additional condition (for example, only for specific model types) or moved to a different part of the eval path. @SunMarc

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2927,6 +2927,20 @@ def prediction_step(
...
+            and getattr(self.args, "use_liger_kernel", False)
+            and inputs.get("labels") is not None
+            and "skip_logits" not in inputs
+        ):
+            try:
+                forward_sig = inspect.signature(unwrap_model(model).forward)
+                if "skip_logits" in forward_sig.parameters:
+                    inputs["skip_logits"] = True
+            except (TypeError, ValueError):
+                pass
```
Comment (SunMarc) [inline src/transformers/trainer.py:2942]: why try except ? in which cases we hit ValueError or TypeError ?

File: tests/trainer/test_skip_logits_eval.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import tempfile
```
Comment (SunMarc) [inline tests/trainer/test_skip_logits_eval.py:1]: don't create a new file, put it in an existing file. Also add one test only, it should be enough. Try to make the tests as simple as possible + small if possible

**Review by SunMarc — COMMENTED** [review]
Thanks, just a few nits

File: src/transformers/trainer.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/trainer.py:2942]: Also this is an arg from liger, so it should always be there no ? or not ?

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -2927,6 +2927,20 @@ def prediction_step(
         else:
             labels = None
 
+        # Enable Liger fused loss path during eval when we only need the loss (no logits).
+        if (
+            prediction_loss_only
+            and getattr(self.args, "use_liger_kernel", False)
```
Comment (SunMarc) [inline src/transformers/trainer.py:2933]: we don't need the getattr

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -2927,6 +2927,20 @@ def prediction_step(
...
+            and inputs.get("labels") is not None
```
Comment (SunMarc) [inline src/transformers/trainer.py:2934]: like here, can can just put this piece of code in the correct place so that we don't have to check that https://github.com/huggingface/transformers/pull/45273/changes

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -2927,6 +2927,20 @@ def prediction_step(
...
+            and "skip_logits" not in inputs
+        ):
```
Comment (SunMarc) [inline src/transformers/trainer.py:2936]: not sure why `skip_logits` will be in inputs
